R2Northstar/NorthstarLauncher
# Mod download UI integration

UI integration of the mod downloading feature merged in https://github.com/R2Northstar/NorthstarLauncher/pull/545

When downloading a mod, this will display download progress, then extraction progress onscreen. Once mod has been downloaded, mods will be reloaded, and server join will pursue.

_**NB: `NSReloadMods` currently doesn't support all mods (weapon skins for instance); some maps are supported.**_

Mods PR: https://github.com/R2Northstar/NorthstarMods/pull/761

### Test instructions

* Install launcher PR
* Install mods PR
* Enable mod downloading ~~in Controls menu~~ **(this is now done by switching the `allow_mod_auto_download` convar to `1`)**
* Join a server ("Space battle" test server should up for you to try this feature, if not ping me please)

### Media

Here's a demo of end-to-end custom map download:

https://github.com/R2Northstar/NorthstarLauncher/assets/11993538/99b61ebd-4dd3-4816-ba31-f6f9137b78fa


### Changes

* Expose download function to squirrel VM;
* `modState` variable is updated regarding mod download state (downloading, extracting etc);
* `NSGetModInstallState` function returns mod download state as a `ModInstallState` struct (defined in squirrel);
* `ModFetchingProgressCallback` function is used to get curl download progress information;
* `GetModArchiveSize` function gives information for extraction progress stats;
* `FetchModSync` isn't run in a thread anymore (since the whole downloading process is already run in a separate thread).
